ThunkLibs/libX11/libX11_Guest.cpp

```diff
@@ -20,134 +20,216 @@ extern "C" {
 #undef max
 }
 
+#include <algorithm>
+#include <array>
+#include <cassert>
 #include <cstdint>
 #include <stdio.h>
 #include <cstring>
 #include <map>
+#include <list>
 #include <string>
+#include <unistd.h>
 
 #include "common/Guest.h"
 #include <stdarg.h>
 
 #include "thunkgen_guest_libX11.inl"
+#include "X11Common.h"
 
 // Custom implementations //
 
 #include <vector>
 
+namespace {
+  // The various X11 variadic functions take a flattened sequence of key:value pairs as arguments.
+  // The key element is a pointer to a string.
+  // The value element is a pointer to the data of that key type.
+  //
+  // Some keys describe function callbacks for various events that the server interface can call.
+  // FEX needs to walk these keys and ensure any callback function has a trampoline so the native
+  // X11 library can call it.
+  template<typename CallbackType>
+  static std::list<CallbackType> ConvertCallbackArguments(std::vector<void*> &IncomingArguments) {
```

Comment:
Much of the logic in this function can be cleaned up by changing the type of this to `std::vector<std::pair<const char*, void*>>`

Reply:
Converting this to a vector of pairs introduces a questionable crash.
In addition it adds some bad castings for the `XGet*` handlers which don't use pairs.

Reply:
> Converting this to a vector of pairs introduces a questionable crash.

What makes it questionable?

> In addition it adds some bad castings for the XGet* handlers which don't use pairs.

How does the current implementation of the function work with these functions? Also, the comment at the top seems very misleading if these functions don't in fact use pairs.

Reply:
The `XGet*` variants push keys in to a `std::vector<void*>` until they hit a nullptr, then passes that array directly through and is handled in the variadic case like previously. It may or may not be correct, which wasn't the point of this PR.

The questionable crash is that with pairs and then casting the to the variadic handler of void* it crashes. Which will be difficult to nail down.

Reply:
Further clarified offline: Changing to `std::vector<std::pair<...>>` would benefit the guest side in `ConvertCallbackArguments`, but the host side would have to do a bit of a dance since `XGet*` doesn't use pairs. The huge 64-value switch cases would have to be slightly different for each.